### PR TITLE
wiremock changes detecting interceptor

### PR DESCRIPTION
**Pull request by spetrychenko**

### CODE REVIEW

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+            String body = getResponseBody(response);
+            return analyzePayloadStructure(resolver, responseBody, body, PayloadType.RESPONSE);
+        } catch (JsonProcessingException e) {
+            log.error("Can't retrieve response body", e);
+        }
+        return Optional.empty();
+    }
+
+    @SuppressWarnings("unchecked")
+    private Optional<String> analyzePayloadStructure(WiremockFileResolver resolver, String payloadBody, String body, PayloadType type) {
```
Comment (artemsnisarenko) [inline xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java:111]: Have you considered using some libraries to do the diffing? E.g. xmlunit/jsonunit that wiremock itself is using. So that the comparison is performed beyond json keys. I'm not certain this method can even handle a json array.
Also manually finding the difference to analyse a warning will be quite labor intensive. Would be nice to have it already available in the log.

File: xs2a-adapter-service-impl/src/test/java/de/adorsys/xs2a/adapter/impl/http/wiremock/FlatMapUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+import java.util.Map;
+
+class FlatMapUtilsTest {
+
+    @Test
+    void flatten() throws JsonProcessingException {
+        String json = "{\"a\":1, \"b\":{\"b1\":2,\"b2\":3},\"c\":[{\"d\":4},5]}";
+        Map map = FlatMapUtils.flatten(new ObjectMapper().readValue(json, Map.class));
+        System.out.println("map = " + map.keySet());
```
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/test/java/de/adorsys/xs2a/adapter/impl/http/wiremock/FlatMapUtilsTest.java]: fixed

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+    @Override
+    public Request.Builder apply(Request.Builder builder) {
+        return builder;
+    }
+
+    @Override
+    public <T> Response<T> postHandle(Request.Builder builder, Response<T> response) {
+        try {
+            WiremockFileResolver fileResolver = WiremockFileResolver.resolve(URI.create(builder.uri()).getPath(), builder.method(), builder.body());
```
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java:66]: what is your proposal? to rename enum or method?

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        }
+        return Optional.empty();
+    }
+
+    private Optional<String> analyzeRequestBody(WiremockFileResolver resolver, Request.Builder builder, String requestBody) {
+        return analyzePayloadStructure(resolver, requestBody, builder.body(), PayloadType.REQUEST);
+    }
+
+    private Optional<String> analyzeRequestHeaders(WiremockFileResolver resolver, Request.Builder builder, Map<String, Object> requestHeaders) {
+        if (requestHeaders != null && !builder.headers().keySet().containsAll(requestHeaders.keySet())) {
```
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java:168]: Ok, what the strategy should be here? To define the list of non-business headers and remove them from comparison?
This strategy could be moved to a separate sub-task.

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java (hunk 1 above)
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java:111]: Of course, I've looked into those libraries but as I understand they do not have the possibility to compare only the structure and I've found this solution on StackOverflow. Write differences into log

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+            log.warn("{} stub headers are different from the request", aspsp.getName());
+            String changes = aspsp.getName() + ":" + resolver.name() + ":request-headers";
+            return Optional.of(changes);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<String> analyzeRequestUrl(WiremockFileResolver resolver, Request.Builder builder, String requestUrl) {
```
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java]: We removed URL comparison because URL has a lot of unique path parameters but if you think that such comparison will be useful we can create a sub-task for that functionality.

File: xs2a-adapter-service-api/src/main/java/de/adorsys/xs2a/adapter/api/ResponseHeaders.java [hunk 6]
```diff
@@ -7,13 +7,15 @@
     public static final String X_REQUEST_ID = "X-Request-ID";
     public static final String ASPSP_SCA_APPROACH = "ASPSP-SCA-Approach";
     public static final String CONTENT_TYPE = "Content-Type";
+    public static final String X_ASPSP_CHANGES_DETECTED = "X-ASPSP-CHANGES-DETECTED";
```
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-api/src/main/java/de/adorsys/xs2a/adapter/api/ResponseHeaders.java]: agree. fixed

File: xs2a-adapter-service-api/src/main/java/de/adorsys/xs2a/adapter/api/http/Request.java [hunk 7]
```diff
@@ -61,6 +72,13 @@
 
         @FunctionalInterface
         interface Interceptor extends UnaryOperator<Builder> {
```
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-api/src/main/java/de/adorsys/xs2a/adapter/api/http/Request.java]: moved

File: xs2a-adapter-service-api/src/main/resources/adapter.config.properties [hunk 8]
```diff
@@ -1,3 +1,7 @@
+# Wiremock interceptor
+
+wiremock.interceptor.enabled=true
```
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-api/src/main/resources/adapter.config.properties]: agree. set to `false`

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/RequestBuilderImpl.java [hunk 9]
```diff
@@ -181,8 +181,21 @@ public boolean emptyBody() {
...
+    public <T> Response<T> send(HttpClient.ResponseHandler<T> responseHandler, Interceptor... interceptors) {
+        for (Interceptor interceptor : interceptors) {
+            if (interceptor != null) {
+                interceptor.preHandle(this);
+            }
+        }
+
+        Response<T> response = httpClient.send(this, responseHandler);
+
+        for (Interceptor interceptor : interceptors) {
```
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/RequestBuilderImpl.java:194]: added java doc

File: xs2a-adapter-service-impl/src/test/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockFileResolverTest.java [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+    private static final String POST_METHOD = "POST";
+    private static final String PUT_METHOD = "PUT";
+    private static final String GET_METHOD = "GET";
+    private static final String CONSENTS_URI = "/v1/consents";
+    private static final String ACCOUNTS_URI = "/v1/accounts";
+    private static final String PAYMENTS_SCT_URI = "/v1/payments/sepa-credit-transfers";
+    private static final String PAYMENTS_PAIN001_SCT_URI = "/v1/payments/pain.001-sepa-credit-transfers";
+    private static final String PERIODIC_SCT_URI = "/v1/periodic-payments/sepa-credit-transfers";
+    private static final String PERIODIC_PAIN001_CST_URI = "/v1/periodic-payments/pain.001-sepa-credit-transfers";
+    // Flowed data
```
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/test/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockFileResolverTest.java]: fixed

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+    @Override
+    public Request.Builder apply(Request.Builder builder) {
+        return builder;
+    }
+
+    @Override
+    public <T> Response<T> postHandle(Request.Builder builder, Response<T> response) {
+        try {
+            WiremockFileResolver fileResolver = WiremockFileResolver.resolve(URI.create(builder.uri()).getPath(), builder.method(), builder.body());
+            String fileName = buildStubFilePath(aspsp.getName(), fileResolver.getFileName());
```
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java]: fixed. thanks

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+            return Optional.empty();
+        }
+        List<Map<String, Object>> bodyMap = (List<Map<String, Object>>) (request).get(BODY_PATTERNS);
+        return Optional.ofNullable((String) bodyMap.get(0).get("equalToJson"));
+    }
+
+    @SuppressWarnings("unchecked")
+    private Optional<Map<String, Object>> getStubRequestHeaders(Map<String, Object> jsonFile) {
+        Map<String, Object> request = (Map<String, Object>) jsonFile.get(REQUEST);
+        if (!request.containsKey(HEADERS)) {
```
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java]: removed unnecessary check

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+        }
+        return Optional.empty();
+    }
+
+    @SuppressWarnings("unchecked")
+    private Optional<String> analyzePayloadStructure(WiremockFileResolver resolver, String payloadBody, String body, PayloadType type) {
+        String payloadType = type.name().toLowerCase();
+        try {
+            Map<String, Object> stubBody = objectMapper.readValue(payloadBody, Map.class);
+            Map<String, Object> currentBody = objectMapper.readValue(body, Map.class);
```
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java:117]: added check on json content type

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+        return builder;
+    }
+
+    @Override
+    public <T> Response<T> postHandle(Request.Builder builder, Response<T> response) {
+        try {
+            WiremockFileResolver fileResolver = WiremockFileResolver.resolve(URI.create(builder.uri()).getPath(), builder.method(), builder.body());
+            String fileName = buildStubFilePath(aspsp.getName(), fileResolver.getFileName());
+            Map<String, Object> jsonFile = readStubFile(fileName);
+            List<String> changes = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java:70]: renamed

File: xs2a-adapter-service-impl/src/test/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptorTest.java [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+        assertThat(actualResponse.getHeaders().getHeadersMap())
+            .containsKey(ResponseHeaders.X_ASPSP_CHANGES_DETECTED)
+            .extractingByKey(ResponseHeaders.X_ASPSP_CHANGES_DETECTED)
+            .matches(val -> val.contains(RESPONSE_PAYLOAD_VALUE))
+            .matches(val -> !val.contains(REQUEST_PAYLOAD_VALUE))
+            .matches(val -> !val.contains(REQUEST_HEADERS_VALUE));
+    }
+
+    @Test
+    void postHandle_nothingMatches() throws JsonProcessingException {
```
[the remark this answers is not in the document]
Comment (berezkin88) [reply xs2a-adapter-service-impl/src/test/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptorTest.java:113]: bug found, working on that, thx

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (artemsnisarenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java:117]: unit tests for request and response bodies in xml would be nice

[the remark this answers is not in the document]
Comment (artemsnisarenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java:117]: you don't plan to support xml at all? silently ignoring it doesn't look like a great handling strategy to me

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java (hunk 1 above)
Comment (artemsnisarenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java]: I don't understand your answer. You're saying that a library makes a more thorough comparison and because of this you decided not to use it.

It looks like what you log as "differences" is just a body not a diff.

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (artemsnisarenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java:176]: some technical headers are also important e.g. content-type:application/json in case of consors

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (artemsnisarenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java]: I think it is important

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java:117]: I've created a separate ticket for that case https://jira.adorsys.de/browse/XS2AAD-707

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (artemsnisarenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java:67]: rename returned type

btw the casing (Wiremock/WireMock) is different from existing files in the same package
https://github.com/adorsys/xs2a-adapter/tree/spe/wiremock/xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java]: The sub-task was created https://jira.adorsys.de/browse/XS2AAD-708

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java:176]: The sub-task was created https://jira.adorsys.de/browse/XS2AAD-709

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java (hunk 1 above)
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java]: For this functionality, we need to compare only the structure of a payload without the field values. But those libraries can't do that. That's why this functionality was written by ourselves.
Fixed logging object, thanks.

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java:67]: renamed

[the remark this answers is not in the document]
Comment (artemsnisarenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java:67]: did you forget to push the changes?

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java (hunk 1 above)
Comment (artemsnisarenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java]: I assume by "this functionality" you mean validation of wiremock mappings. Comparison of field values (types/sets of acceptable values) is necessary for that.

The logged object is still not a diff.

File: xs2a-adapter-service-impl/src/test/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptorTest.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (berezkin88) [reply xs2a-adapter-service-impl/src/test/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptorTest.java:132]: fixed

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java]: oops, i did. pushed

File: xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java (hunk 1 above)
Comment (spetrychenko) [reply xs2a-adapter-service-impl/src/main/java/de/adorsys/xs2a/adapter/impl/http/wiremock/WiremockStubDifferenceDetectingInterceptor.java]: 3rd try.

**Review by berezkin88 — APPROVED** [review]
